### ISSUE TITLE
[codebuild] Handle CodeBuild "WINDOWS_SERVER_2019_CONTAINER" environment type for Windows Server 2019 Docker images

### ISSUE THREAD

**Issue by ermanno** [aws/aws-cdk#9484]
The [CodeBuild documentation](https://docs.aws.amazon.com/codebuild/latest/userguide/troubleshooting.html#troubleshooting-windows-version-mismatch) indicates that when using a Windows Server 2019 based Docker image the environment type should be `WINDOWS_SERVER_2019_CONTAINER`, otherwise when building there will be an error like:

>BUILD_CONTAINER_UNABLE_TO_PULL_IMAGE: Unable to pull customer's container image. CannotPullContainerError: a Windows version 10.0.17763-based image is incompatible with a 10.0.14393 host

Currently [BuildEnvironment](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-codebuild.BuildEnvironment.html) doesn't support specifying the environment type, and [IBuildImage](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-codebuild.IBuildImage.html) supports only retrieving the environment type, not setting it.

### Use Case

CDK users that have Windows Server 2019 based custom images which they retrieve with methods such as `WindowsBuildImage.fromDockerRegistry()` would end up with CodeBuild failures like the one indicated above, since the environment type would be set to `WINDOWS_CONTAINER` rather than to `WINDOWS_SERVER_2019_CONTAINER`.

This is similar to some other issues such as:
1. `LINUX_GPU_CONTAINER` environment type: https://github.com/aws/aws-cdk/issues/8408
1. `ARM_CONTAINER` environment type: https://github.com/aws/aws-cdk/pull/5541

### Proposed Solution

As in the issues above we could create a new class of `IBuildImage`.

* [X] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

**Comment by skinny85** [comment]
Relevant docs: https://docs.aws.amazon.com/codebuild/latest/userguide/build-env-ref-compute-types.html

**Comment by ermanno** [comment]
I have made some changes based on what I have seen in [the ARM patch](https://github.com/aws/aws-cdk/pull/5541/files). I have some doubts and wanted to solicit feedback regarding some points:
1. I have tried to factor out code common to the `WindowsBuildImage` class, I was unsure about how we wanted to handle `private constructor(props: WindowsBuildImageProps)`
1. I was unsure about how to go about testing this code, in particular I see that for the other similar patches I found that what was tested was the constants images. I only found one test mentioning the WINDOWS_CONTAINER (for the [WindowsBuildImage.WINDOWS_BASE_2_0](https://github.com/aws/aws-cdk/blob/16d2a48a4d0e049d0ae6cd8c9165eccc816ebd4b/packages/%40aws-cdk/aws-codebuild/test/test.codebuild.ts#L513) constant).
